// @sap/cds-dk/lib/init/template/workzone-standard/index.js
const cds = require('../../../cds')
const { readProject } = require('../../projectReader')
const { merge } = require('../../merge')
const { join } = require('path')
const { destination, destinations, srv4 } = require('../../registries/mta')
const { odataApi } = require('../../registries/xs-app')

module.exports = class extends require('../../plugin') {

    static help() {
      return 'SAP BTP Work Zone, Standard Edition'
    }

    requires() {
      return ['html5-repo', 'destination']
    }

    static hasInProduction(env) {
      return !!env.requires?.workzone
    }

    canRun() {
      if (cds.requires.multitenancy) {
        throw `SAP BTP Work Zone, Standard Edition does not support multitenancy yet.`
      }
      return true
    }

    async run() {
      const project = readProject()
      const { appUIPaths, appPath } = project
      await Promise.all(appUIPaths.map(p => {
        project.uiApp = p
        return merge(__dirname, 'files/manifest.json.hbs').into(join(appPath, p, 'webapp/manifest.json'), { project })
      }))
      await merge({ cds: { requires: { workzone: true } } }).into('package.json')
    }

    async combine() {
      const project = readProject()
      const { hasMta, srvPath, apps, appPath } = project

      await Promise.all(apps.map(async ({app}) => {
        const additions = [odataApi]
        const overwrites = [{
          item: [odataApi, 'destination'],
          withValue: `${project.appName}-srv-api`
        }]
        project.app = app
        await merge(__dirname, 'files/xs-app.json.hbs').into(join(appPath, app, 'xs-app.json'), { with: project, additions, overwrites })
      }))

      if (hasMta) {
        const srv = srv4(srvPath)
        const additions = [srv, destinations, destination,
          {
            in: [destination, 'requires'],
            where: { name: 'srv-api' }
          }
        ]
        await merge(__dirname, 'files/mta.yaml.hbs').into('mta.yaml', { with: project, additions })
      }
  }
}
