// @sap/eslint-plugin-cds/lib/rules/auth-valid-restrict-where.js
'use strict'

const cds = require('@sap/cds')

module.exports = {
  meta: {
    schema: [{/* to avoid deprecation warning for ESLint 9 */}],
    docs: {
      description: '`@restrict.where` must have valid values.',
      category: 'Model Validation',
      recommended: true,
      url: 'https://cap.cloud.sap/docs/tools/cds-lint/meta/auth-valid-restrict-where',
    },
    severity: 'error',
    messages: {
      CompilationFailed: 'Invalid `where` expression, CDS compilation failed.',
    },
    type: 'problem',
    model: 'inferred'
  },
  create (context) {
    const model = context.getModel()
    if (!model)
      return

    return function checkRestrictWhere() {
      model.foreach('entity', checkEntityRestrict)
    }

    function checkEntityRestrict (def) {
      if (!Array.isArray(def['@restrict']))
        return

      for (const entry of def['@restrict']) {
        if (entry.where) {
          // TODO: Check return value (where xpr)
          checkAndCompileWhereExpression(entry.where)
        }
      }

      function checkAndCompileWhereExpression(where) {
        try {
          const compileOptions = {}
          return typeof where === 'string'
            ? cds.parse.expr(where, compileOptions)
            : where

        } catch {
          context.report({
            messageId: 'CompilationFailed',
            node: context.getNode(def),
            file: def.$location.file,
          })
          return null
        }
      }
    }
  }
}
